sklearn/random_projection.py

```diff
@@ -387,7 +389,9 @@ def fit(self, X, y=None):
             self.n_components_ = self.n_components
 
         # Generate a projection matrix of size [n_components, n_features]
-        self.components_ = self._make_random_matrix(self.n_components_, n_features)
+        self.components_ = self._make_random_matrix(
+            self.n_components_, n_features
+        ).astype(X.dtype)
```

Comment:
We can prevent a copy for the `np.float64` case:

```suggestion
        ).astype(X.dtype, copy=False)
```

Although, there is still a memory copy for float32, which is technically a regression for the float32 case.

The "best solution" would be to use [Generator.standard_normal](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.standard_normal.html#numpy.random.Generator.standard_normal), which supports different dtype outputs. This way we do not need  casting. To go down this path, we would need to support Generators, which is a long term goal.

In the short term, we need to be content with this memory copy regression for float32. Practically, `n_components by n_features` matrix is not too big, so I think it should be okay. I would still wait to see what others think.

Reply:
> In the short term, we need to be content with this memory copy regression for float32. Practically, n_components by n_features matrix is not too big, so I think it should be okay. I would still wait to see what others think.

I agree we can probably leave with it. The non-copy of X with shape `(n_samples, n_features)` is likely to yield a net memory efficiency improvement.